src/networkapi.jl

```diff
@@ -1481,5 +1481,158 @@ function validate(rs::ReactionSystem, info::String = "")
+# """
+# 
+# function rateweightedgraph(rs::ReactionSystem, rates::Dict{Any, Float64}) 
+#     if length(rates) != numparams(rs) 
+#         error("The number of reaction rates must be equal to the number of parameters")
+#     end
+# 
+#     complexes, D = reactioncomplexes(rs)
+#     rxns = reactions(rs)
+# 
+#     g = incidencematgraph(rs)
+#     rwg = MetaDiGraph(g)
+# 
+#     for v in vertices(rwg)
+#         set_prop!(rwg, v, :complex, complexes[v])
+#     end
+# 
+#     for (i, e) in collect(enumerate(edges(rwg)))
+#         rxn = rxns[i]
+#         set_prop!(rwg, Graphs.src(e), Graphs.dst(e), :reaction, rxn)
+#         set_prop!(rwg, Graphs.src(e), Graphs.dst(e), :rate, rates[rxn.rate])
+#     end
+# 
+#     rwg
+# end
+
+function ratematrix(rs::ReactionSystem, rates::Dict{Any, Float64}) 
+    if length(rates) != numparams(rs) 
+        error("The number of reaction rates must be equal to the number of parameters")
+    end
+
+    complexes, D = reactioncomplexes(rs)
+    n = length(complexes)
+    rxns = reactions(rs)
+    ratematrix = zeros(n, n)
+
+    for r in 1:length(rxns)
+        rxn = rxns[r]
+        s = findfirst(x->x==-1, D[:,r])
+        p = findfirst(x->x==1, D[:,r])
+        ratematrix[s, p] = rates[rxn.rate]
+    end
+    ratematrix
+end
+
+"""
+"""
+function matrixtree(g::SimpleDiGraph, distmx::Matrix)
+    n = nv(g)
+    if size(distmx) != (n, n)
+        error("Size of distance matrix is incorrect")
+    end
+
+    π = zeros(n) 
+
+    if !Graphs.is_connected(g)
+        ccs = Graphs.connected_components(g)
+        for cc in ccs
+            sg, vmap = Graphs.induced_subgraph(g, cc)
+            distmx_s = distmx[cc, cc]
+            π_j = matrixtree(sg, distmx_s)
+            π[cc] = π_j
+        end
+        return π
+    end
+
+    # generate all spanning trees
+    ug = SimpleGraph(SimpleDiGraph(g))
+    trees = collect(Combinatorics.combinations(collect(edges(ug)), n-1))
+    trees = SimpleGraph.(trees)
+    trees = filter!(t->isempty(Graphs.cycle_basis(t)), trees)
+    # trees = spanningtrees(g)
+
+    # constructed rooted trees for every edge, compute sum
+    for v in 1:n
+        rootedTrees = [reverse(Graphs.bfs_tree(t, v, dir=:in)) for t in trees]
+        π[v] = sum([treeweight(t, g, distmx) for t in rootedTrees])
+    end
+    
+    # sum the contributions
+    return π 
+end
+
+function treeweight(t::SimpleDiGraph, g::SimpleDiGraph, distmx::Matrix) 
+    prod = 1
+    for e in edges(t)
+        s = Graphs.src(e); t = Graphs.dst(e)
+        prod *= distmx[s, t] 
+    end
+    prod 
+end
+
+# TODO: implement Winter's algorithm for generating spanning trees 
+function spanningtrees(g::SimpleGraph) 
+        
+end
+
 # Checks if a unit consist of exponents with base 1 (and is this unitless).
-unitless_exp(u) = istree(u) && (operation(u) == ^) && (arguments(u)[1] == 1)
\ No newline at end of file
+unitless_exp(u) = istree(u) && (operation(u) == ^) && (arguments(u)[1] == 1)
```

Comment:
This should go before your added code since it goes with the unit validation above.